### PR TITLE
[#53] Added Search Bar functionality on the Assignments page

### PR DESCRIPTION
**Pull request by saksham1919**
Added code to enable Search Bar functionality on the Assignments page.

## Description
The system has been improved with a search bar on the assignments page, allowing users to easily filter and search for their specific requirements.

## How has this been tested?
RSpec and Cucumber tests have been successfully completed for all scenarios, with all tests passing. The functionality has also been verified locally, and the screenshots below provide evidence of this.

## Screenshots (if appropriate):
Search-bar implementation:

image
image
&nbsp;
Coverage report (after running rspec and cucumber acceptance tests in tandem):

image

## Checklist:

- [*] Zero `rubocop` violations
- [*] My code receives an "A" when `rubycritic` is run
- [*] >90% test coverage of my changes
- [*] All new and existing tests passed

### CODE REVIEW

**Review by masarkar-tamu — DISMISSED** [review]

**Review by Banna150149 — DISMISSED** [review]

File: app/controllers/assignments_controller.rb [hunk 1]
```diff
@@ -24,6 +24,7 @@ def edit
 
   # POST /assignments or /assignments.json
   def create
+    assignment_params[:repository_name].downcase!
```
Comment (maxsmith271346) [inline app/controllers/assignments_controller.rb:27]: Maybe make this a `before_validation` callback, and maybe use this instead:
```
def generate_github_repo_name(input)
  repo_name = input.downcase
  
  # replace spaces/underscores with hyphens
  repo_name.gsub!(/[ _]/, '-')
  
  # remove any non-alphanumeric/hyphen chars
  repo_name.gsub!(/[^a-z0-9\-]/, '')
  
  # remove leading/trailing hyphens
  repo_name.gsub!(/^-+|-+$/, '')

  repo_name
end
```

File: app/controllers/assignments_controller.rb [hunk 2]
```diff
@@ -40,9 +41,10 @@ def create
     end
   end
 
-  # PATCH/PUT /assignments/1 or /assignments/1.json
   # PATCH/PUT /assignments/1 or /assignments/1.json
   def update
+    assignment_params[:repository_name].downcase!
```
Comment (maxsmith271346) [inline app/controllers/assignments_controller.rb:46]: See above

File: app/javascript/controllers/search_controller.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import { Controller } from "@hotwired/stimulus";
+
+// Connects to data-controller="search"
+export default class extends Controller {
+  static targets = ["input", "clearButton"];
+
+  // Trigger search on Enter key press
+  handleEnter(event) {
+    if (event.key === "Enter") {
```
Comment (maxsmith271346) [inline app/javascript/controllers/search_controller.js:9]: Why is there an empty if-statement here?

File: features/step_definitions/search_bar_steps.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
+  Given('We have the following assignments exist:') do |table|
```
Comment (maxsmith271346) [inline features/step_definitions/search_bar_steps.rb:1]: This is duplicated from another step definition. Please remove.

File: features/step_definitions/search_bar_steps.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    table.hashes.each do |assignment|
+      Assignment.create!(
+        assignment_name: assignment['assignment_name'],
+        repository_name: assignment['repository_name']
+      )
+    end
+  end
+
+  Given('I am on the assignments page') do
```
Comment (maxsmith271346) [inline features/step_definitions/search_bar_steps.rb:10]: This also looks like a potential duplicate

File: features/step_definitions/search_bar_steps.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    visit assignments_path
+  end
+
+  When('I enter {string} into the search bar') do |query|
+    fill_in 'query', with: query
+  end
+
+  When('I click the search button') do
```
Comment (maxsmith271346) [inline features/step_definitions/search_bar_steps.rb:18]: So is this

File: features/step_definitions/search_bar_steps.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        repository_name: assignment['repository_name']
+      )
+    end
+  end
+
+  Given('I am on the assignments page') do
+    visit assignments_path
+  end
+
+  When('I enter {string} into the search bar') do |query|
```
Comment (maxsmith271346) [inline features/step_definitions/search_bar_steps.rb:14]: So is this

**Review by maxsmith271346 — CHANGES_REQUESTED** [review]
Please submit a screenshot of the simplecov report

File: features/step_definitions/search_bar_steps.rb (hunk 4 above)
Comment (maxsmith271346) [reply features/step_definitions/search_bar_steps.rb:1]: This does the same thing as `"the following assignments exist:"`

File: features/step_definitions/search_bar_steps.rb (hunk 5 above)
Comment (maxsmith271346) [reply features/step_definitions/search_bar_steps.rb:10]: Same as `'I am on the {string} page'`

File: features/step_definitions/search_bar_steps.rb (hunk 6 above)
Comment (maxsmith271346) [reply features/step_definitions/search_bar_steps.rb:18]: Same as `'I click the {string} button'`

File: features/step_definitions/search_bar_steps.rb (hunk 7 above)
Comment (maxsmith271346) [reply features/step_definitions/search_bar_steps.rb:14]: Seems to be same as `'I fill in {string} with {string}'`

File: app/controllers/assignments_controller.rb (hunk 1 above)
Comment (masarkar-tamu) [reply app/controllers/assignments_controller.rb:27]: Thanks for the comment! Implemented this, by providing a `before_validation` hook in the Assignment model.

However, I had to make a small adjustment in the controller. In the `AssignmentController#create` the `assignment_repo_init` function was called before the `@assignment.save` statement (where validation hooks are called).

As a result, the validation hook was called after the repo was created in our organization, which was causing a few errors for the cases of incorrect repository name entries (for example, those that have whitespaces, etc. ). 

So, I shifted the `assignment_repo_init` call slightly after the `@assignment.save' statement. This allows the repo to be created after the validation hooks are called. (See below)

```
  def create
    @assignment = Assignment.new(assignment_params)
    github_token = session[:github_token]

    respond_to do |format|
      if @assignment.save
        @assignment.assignment_repo_init(github_token)
        format.html { redirect_to @assignment, notice: "Assignment was successfully created." }
        format.json { render :show, status: :created, location: @assignment }
      else
        format.html { render :new, status: :unprocessable_entity }
        format.json { render json: @assignment.errors, status: :unprocessable_entity }
      end
    end
  end

File: app/controllers/assignments_controller.rb (hunk 2 above)
Comment (masarkar-tamu) [reply app/controllers/assignments_controller.rb:46]: Solved, thanks!

File: app/javascript/controllers/search_controller.js (hunk 3 above)
Comment (masarkar-tamu) [reply app/javascript/controllers/search_controller.js:9]: Good observation, Max, and thank you for pointing it out. 

This line was not necessary, as I have recently found out that in HTML, a form gets "submitted" automatically when someone presses Enter after typing their query, and no additional event listeners are needed.

Accordingly, I have removed the `handleEnter` function and I have verified that the overall functionality of the search-bar has remained the same.

File: features/step_definitions/search_bar_steps.rb (hunk 7 above)
Comment (saksham1919) [reply features/step_definitions/search_bar_steps.rb]: That's right! Modified it accordingly.

File: features/step_definitions/search_bar_steps.rb (hunk 4 above)
Comment (saksham1919) [reply features/step_definitions/search_bar_steps.rb]: Right, we can reuse the existing step definiton

File: features/step_definitions/search_bar_steps.rb (hunk 5 above)
Comment (saksham1919) [reply features/step_definitions/search_bar_steps.rb]: To reuse the existing step definition, it needs to be extended. To address this, I have updated the step definition to also support the assignments page, as it previously only handled the login page.

File: features/step_definitions/search_bar_steps.rb (hunk 6 above)
Comment (saksham1919) [reply features/step_definitions/search_bar_steps.rb]: done.

**Review by maxsmith271346 — APPROVED** [review]
I'm approving this, since we have >90% overall test coverage and your feature is adequately tested, but individually the RSpec & Cucumber are each under 90%.  I think that some of this is due to improper use of mocking in a lot of the Sprint 1 tests.

File: features/step_definitions/search_bar_steps.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
+  Given('I have searched for {string}') do |query|
```
Comment (maxsmith271346) [inline features/step_definitions/search_bar_steps.rb:1]: Please remove the redundant, unused step definitions

**Review by maxsmith271346 — CHANGES_REQUESTED** [review]

File: features/step_definitions/search_bar_steps.rb (hunk 8 above)
Comment (saksham1919) [reply features/step_definitions/search_bar_steps.rb:1]: This step definition is actually being used. Please refer to Scenario 3 in search_bar.feature file.

Comment (maxsmith271346) [reply features/step_definitions/search_bar_steps.rb:1]: Oh sorry, I missed that.  Could we replace the code in there with step definitions then?
```
Given(...) do |query|
  steps %(
    Given I file in "query" with #{query}
    And I click the "Search Assignment" button
  )

Comment (saksham1919) [reply features/step_definitions/search_bar_steps.rb]: Yeah, we can do that.

**Review by maxsmith271346 — APPROVED** [review]

**Pull request merged**